app/workers/work-reaper.ts

```diff
@@ -25,48 +25,99 @@ export default class WorkReaper implements Worker {
+        );
+        if (workItemIds.length > 0) {
+          const numItemsDeleted = await deleteWorkItemsById(db, workItemIds);
+          totalDeleted += numItemsDeleted;
+          this.logger.info(`Work reaper removed ${numItemsDeleted} work items, starting id: ${startingId}.`);
+          startingId = Math.max(...workItemIds);
+        } else {
+          this.logger.info('Work reaper did not find any work items to delete');
+        }
+
+        if (workItemIds.length < batchSize) {
+          done = true;
         }
-      } else {
-        this.logger.info('Work reaper did not find any work items to delete');
+      } catch (e) {
+        this.logger.error('Error attempting to delete terminal work items');
+        this.logger.error(e);
+        done = true;
       }
-      const workStepIds = await getWorkflowStepIdsByJobUpdateAgeAndStatus(
-        db, notUpdatedForMinutes, jobStatus,
-      );
-      if (workStepIds.length) {
-        const chunkedWorkStepIds = _.chunk(workStepIds, batchSize);
-        for (const workStepIdsChunk of chunkedWorkStepIds) {
-          const numItemsDeleted = await deleteWorkflowStepsById(db, workStepIdsChunk);
-          this.logger.info(`Work reaper removed ${numItemsDeleted} workflow steps`);
+    }
+    this.logger.info(`Work reaper delete terminal work items completed. Total work items deleted: ${totalDeleted}`);
+  }
+
+  /**
+   * Find workflow steps that are older than notUpdatedForMinutes and delete them.
+   * @param notUpdatedForMinutes - upper limit on the duration since the last update
+   * @param jobStatus - a list of terminal job statuses
+   * @returns Resolves when the request is complete
+   */
+  async deleteTerminalWorkflowSteps(notUpdatedForMinutes: number, jobStatus: JobStatus[]): Promise<void> {
+    let done = false;
+    let startingId = 0;
+    let totalDeleted = 0;
+    const batchSize = env.workReaperBatchSize;
+    this.logger.info('Work reaper delete terminal workflow steps started.');
+
+    while (!done) {
+      try {
+        const workflowSteps = await getWorkflowStepIdsByJobUpdateAgeAndStatus(
+          db, notUpdatedForMinutes, jobStatus, startingId, batchSize,
+        );
+        if (workflowSteps.length > 0) {
+          const numItemsDeleted = await deleteWorkflowStepsById(db, workflowSteps);
+          totalDeleted += numItemsDeleted;
+          this.logger.info(`Work reaper removed ${numItemsDeleted} workflow steps, starting id: ${startingId}.`);
+          startingId = Math.max(...workflowSteps);
+        } else {
+          this.logger.info('Work reaper did not find any workflow steps to delete');
+        }
+
+        if (workflowSteps.length < batchSize) {
+          done = true;
         }
-      } else {
-        this.logger.info('Work reaper did not find any workflow steps to delete');
+      } catch (e) {
+        this.logger.error('Error attempting to delete terminal workflow steps');
+        this.logger.error(e);
+        done = true;
```

Comment:
Can you move the `done = true` up above the logging - just to be sure that an exception with the logging for some reason doesn't cause things to run in a tight infinite loop. Same thing with the exception handling in the other while loop on line 62.